pkg/hds/tracker/callbacks.go

```diff
@@ -0,0 +1,217 @@
+				streams.watchdogCancel()
+			}
+			delete(t.dpStreams, dp)
+		}
+	}
+}
+
+func (t *tracker) OnHealthCheckRequest(streamID xds.StreamID, req *envoy_service_health.HealthCheckRequest) error {
+	t.metrics.RequestsReceivedMetric.Inc()
+
+	id, err := xds.ParseProxyIdFromString(req.GetNode().GetId())
+	if err != nil {
+		t.log.Error(err, "failed to parse Dataplane Id out of HealthCheckRequest", "streamid", streamID, "req", req)
+		return nil
+	}
+
+	dataplaneKey := core_model.ResourceKey{Mesh: id.Mesh, Name: id.Name}
+
+	t.Lock()
+	defer t.Unlock()
+
+	streams := t.dpStreams[dataplaneKey]
+	if streams.activeStreams == nil {
+		streams.activeStreams = map[xds.StreamID]bool{}
+	}
+	streams.activeStreams[streamID] = true
+
+	if streams.watchdogCancel == nil { // watchdog was not started yet
+		stopCh := make(chan struct{})
+		streams.watchdogCancel = func() {
+			close(stopCh)
+		}
+		// kick off watchdog for that Dataplane
+		go t.newWatchdog(req.Node).Start(stopCh)
+		t.log.V(1).Info("started Watchdog for a Dataplane", "streamid", streamID, "proxyId", id, "dataplaneKey", dataplaneKey)
+	}
+	t.dpStreams[dataplaneKey] = streams
+	t.streamsAssociation[streamID] = dataplaneKey
+	return nil
+}
+
+func (t *tracker) newWatchdog(node *envoy_core.Node) watchdog.Watchdog {
+	return &watchdog.SimpleWatchdog{
+		NewTicker: func() *time.Ticker {
+			return time.NewTicker(t.config.RefreshInterval)
+		},
+		OnTick: func() error {
+			start := core.Now()
+			defer func() {
+				t.metrics.HdsGenerations.Observe(float64(core.Now().Sub(start).Milliseconds()))
+			}()
+			return t.reconciler.Reconcile(node)
+		},
+		OnError: func(err error) {
+			t.metrics.HdsGenerationsErrors.Inc()
+			t.log.Error(err, "OnTick() failed")
+		},
+		OnStop: func() {
+			if err := t.reconciler.Clear(node); err != nil {
+				t.log.Error(err, "OnTick() failed")
+			}
+		},
+	}
+}
+
+func (t *tracker) OnEndpointHealthResponse(streamID xds.StreamID, resp *envoy_service_health.EndpointHealthResponse) error {
+	t.metrics.ResponsesReceivedMetric.Inc()
+	for _, clusterHealth := range resp.GetClusterEndpointsHealth() {
+		if len(clusterHealth.LocalityEndpointsHealth) == 0 {
+			continue
+		}
+		if len(clusterHealth.LocalityEndpointsHealth[0].EndpointsHealth) == 0 {
+			continue
+		}
+		status := clusterHealth.LocalityEndpointsHealth[0].EndpointsHealth[0].HealthStatus
+		health := status == envoy_core.HealthStatus_HEALTHY || status == envoy_core.HealthStatus_UNKNOWN
+		port, err := names.GetPortForLocalClusterName(clusterHealth.ClusterName)
+		if err != nil {
+			return err
+		}
+		if err := t.updateDataplane(streamID, port, health); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (t *tracker) updateDataplane(streamID xds.StreamID, port uint32, ready bool) error {
+	t.RLock()
+	defer t.RUnlock()
+	dataplaneKey, hasAssociation := t.streamsAssociation[streamID]
+	if !hasAssociation {
+		return errors.Errorf("no proxy for streamID = %d", streamID)
+	}
+
```

Comment:
If I understand this code, it means that for every interval (5s) we will be fetching the DP from the database (uncached).
Is there any other place in the code that modifies health? If not, we could store this data here in some map and update only if there is a change. Look at DataplaneStatusSink and DataplaneInsight, it updates only if there is a change in the insight.

Reply:
Users can manually update Dataplane resource and wipe out `health` section. Technically we can add logic into DataplaneManager.Update to set `health`, but this means `callbacks` won't be the only place where we modify health